# <html>
   
<div style="background-image: linear-gradient(to left, rgb(255, 255, 255), rgb(138, 136, 136)); width: auto; margin: 10px;">
  <img src="https://upload.wikimedia.org/wikipedia/en/thumb/f/fd/University_of_Tehran_logo.svg/225px-University_of_Tehran_logo.svg.png" width=100px width=auto style="padding:10px; vertical-align: center;">

</div>
   
<div   style:"text-align: center; background-image: linear-gradient(to left, rgb(255, 255, 255), rgb( 219, 204, 245  ));width: 400px; height: 30px; ">
<h1 style="font-family: Georgia; color: black; text-align: center; ">Course: Computer Vision </h1>

</div>
    <div   style:"border: 3px solid green;text-align: center; ">
<h1 style="font-family: Georgia; color: black; text-align: center; ">Project1 - Q3 </h1>

</div>

   <div>    
<h1 style="font-family: Georgia; color: black; text-align: center; font-size:15px;">Mohammad Amanlou- sid:810100084 </h1>

</div>
   

</html>

**Report on Exercise: Implementation with and without Libraries**

In this report, I have explored both approaches to implement the exercise. First, I completed the implementation using popular computer vision libraries such as OpenCV and Scikit-Image. Then, in order to earn additional credit, I implemented all exercises from scratch without using any pre-built libraries. In this section, I only utilized the NumPy and SciPy libraries and did not use any other libraries.

By doing so, my goal was to provide a comprehensive analysis of two different methods for solving problems related to computer vision.

In [1]:
import os
import numpy as np
import imageio
import struct

img_path = "input_image.jpg"  
img = imageio.imread(img_path)

def get_image_header_info(image_path):
    with open(image_path, 'rb') as f:
        header = f.read(8)  
    return header

def get_image_resolution(image):
    height, width = image.shape[:2]
    return width, height

def check_compression(image_path):
    extension = os.path.splitext(image_path)[1].lower()
    
    if extension == ".jpg" or extension == ".jpeg":
        return "JPEG Compression"
    elif extension == ".png":
        return "PNG Compression (Lossless)"
    elif extension == ".bmp":
        return "BMP Format (Uncompressed)"
    elif extension == ".gif":
        return "GIF Compression (Lossless)"
    else:
        return "Unknown compression or format"

header_info = get_image_header_info(img_path)
resolution = get_image_resolution(img)
compression_info = check_compression(img_path)

print("Image Header Information (First 8 bytes):", header_info)
print(f"Image Resolution: {resolution[0]}x{resolution[1]} (Width x Height)")
print(f"Compression Type: {compression_info}")


Image Header Information (First 8 bytes): b'\xff\xd8\xff\xe0\x00\x10JF'
Image Resolution: 960x1280 (Width x Height)
Compression Type: JPEG Compression


C:\Users\a\AppData\Local\Temp\ipykernel_8392\1113080910.py:7: DeprecationWarning: Starting with ImageIO v3 the behavior of this function will switch to that of iio.v3.imread. To keep the current behavior (and make this warning disappear) use `import imageio.v2 as imageio` or call `imageio.v2.imread` directly.
  img = imageio.imread(img_path)


## Detailed Report on Image Format Analysis 📸

In this report, I describe the process and results of analyzing the input image, which is a color image, in terms of its format, header, bitstream, resolution, compression, and compression algorithm. I implemented this analysis using a Python script that extracts key properties of the image such as its header, resolution, compression type, and bitstream details. Below, I will walk you through the steps I took, the code used, and the final results of the analysis.

### 1. **Image Loading** 📥

I started by loading the image using the `imageio` library. The image was read into a NumPy array, which allows for easy manipulation and extraction of various image properties. The image file was a JPG format image, and I used `imageio.imread()` to load it directly into Python. 

### 2. **Extracting Image Header Information** 🏷️

The first key step was to inspect the **header** of the image. Every image format, such as JPEG, PNG, or BMP, starts with a unique set of bytes known as the **header**. These bytes give us important information about the file format and type. To extract this, I used the following method:

- I opened the image in binary mode (`rb`), and I read the first 8 bytes of the file.
- For the JPG file, the first 8 bytes came out as `b'\xff\xd8\xff\xe0\x00\x10JF'`. This specific byte sequence is indicative of a **JPEG format** file, where:
  - `0xFF 0xD8` marks the start of a JPEG file.
  - `0xFF 0xE0` signifies the start of the APP0 marker, which contains metadata for JPEG files.
  - The remaining bytes `00 10` and `JF` are used to indicate specific JPEG format information.

This header information confirms that the image is indeed in the JPEG format, and these bytes are typical for JPEG files. 📄

### 3. **Resolution Extraction** 📏

Next, I extracted the **resolution** of the image. The resolution provides us with the dimensions of the image in terms of **width** and **height**. To get this, I simply used the `img.shape` attribute, which directly gives the dimensions of the image in pixels. For the loaded image:

- **Width**: 960 pixels
- **Height**: 1280 pixels

This indicates that the image is a vertical portrait style image. Knowing the resolution helps understand the image's physical size and is crucial when it comes to compressing or scaling the image. 🖼️

### 4. **Compression Type** 🗜️

One of the most important aspects to examine is whether the image is **compressed** and if so, what compression method is used. Image compression can be **lossy** (e.g., JPEG) or **lossless** (e.g., PNG). 

For the JPEG image I analyzed, the compression type is clearly **JPEG Compression**. This is a **lossy** compression method, meaning that some image data is lost during the compression process, which reduces the file size but also slightly degrades the image quality. The algorithm works by discarding less important data (such as high-frequency details), leading to significant reductions in file size, especially for images with complex color patterns and gradients. 

The JPEG format is widely used because it balances **compression** and **image quality**, making it suitable for web use, photography, and other applications where file size matters. The JPEG algorithm specifically uses techniques like **DCT (Discrete Cosine Transform)** to compress the image in a way that maximizes efficiency. 🔄

### 5. **Bitstream Analysis** 💾

The **bitstream** of the image refers to the raw binary data that makes up the image file. When we open an image file, we are essentially interacting with its bitstream, which is composed of header information, pixel data, and compression metadata. 

In this case, the bitstream for the JPEG image is encoded in such a way that the raw pixel data is compressed using the **JPEG compression algorithm**. The header contains metadata, including the image size, and the rest of the file holds the compressed pixel data. The compression algorithm reduces the bitstream size by encoding redundant information more efficiently.

To fully analyze the bitstream in detail, one would need to parse the JPEG format thoroughly, but for now, we know that it follows the structure described in the JPEG standard, with a combination of compressed data and some metadata.

---

### Final Results:

From the analysis, I gathered the following insights:

- **Image Format:** The image is a **JPEG** file, as confirmed by the header bytes `b'\xff\xd8\xff\xe0\x00\x10JF'` (JPEG signature).
- **Resolution:** The image's resolution is **960 x 1280** pixels, indicating a portrait orientation.
- **Compression Type:** The image uses **JPEG Compression**. This is a **lossy compression** technique, meaning some quality is sacrificed for a reduced file size. JPEG is widely used due to its balance of image quality and file size.
- **Bitstream:** The bitstream for this JPEG file is compressed using the JPEG algorithm. It contains the necessary metadata in the header and the compressed pixel data, which significantly reduces the file size compared to uncompressed formats like BMP.

